### PR TITLE
measureme: Add a ProfilingDataBuilder that can be used for constructing test data.

### PR DESCRIPTION
**Pull request by michaelwoerister**
This is useful because data generated from an actual `Profiler` contains unpredictable timestamps.

r? @wesleywiser

### CODE REVIEW

**Comment by michaelwoerister** [comment]
I haven't found the time to write actual test cases with this yet (e.g. for `summarize`). Maybe later today.

**Comment by michaelwoerister** [comment]
Beware! `dbg!` strikes again. Let me fix that.

**Comment by michaelwoerister** [comment]
> Beware! `dbg!` strikes again. Let me fix that.

Fixed.

**Comment by michaelwoerister** [comment]
I added a single test case to `summarize` just now. That's all I'm going to do for this PR, so feel free to review any time.

File: measureme/src/profiling_data.rs [hunk 1]
```diff
@@ -21,6 +29,15 @@ impl ProfilingData {
             fs::read(paths.string_index_file).expect("couldn't read string_index file");
         let event_data = fs::read(paths.events_file).expect("couldn't read events file");
 
+        let event_data_format = read_file_header(&event_data, FILE_MAGIC_EVENT_STREAM)?;
+        if event_data_format != CURRENT_FILE_FORMAT_VERSION {
```
Comment (wesleywiser) [inline measureme/src/profiling_data.rs]: Do we need to support `CURRENT_FILE_FORMAT_VERSION - 1` for perf.rlo?

cc @Mark-Simulacrum

File: measureme/src/profiling_data.rs [hunk 2]
```diff
@@ -152,3 +169,234 @@ impl<'a> Iterator for MatchingEventsIterator<'a> {
...
+
+/// A `ProfilingDataBuilder` allows for programmatically building
+/// `ProfilingData` objects. This is useful for writing tests that expect
+/// `ProfilingData` with predictable events (and especially timestamps) in it.
+///
+/// `ProfilingDataBuilder` provides a convenient interface but its
+/// implementation might not be efficient, which why it should only be used for
+/// writing tests and other things that are not performance sensitive.
+pub struct ProfilingDataBuilder {
+    event_sink: Arc<ByteVecSink>,
```
Comment (wesleywiser) [inline measureme/src/profiling_data.rs]: Do we need to use `Arc`s here? Unit tests should be single threaded so we don't need this to work across multiple threads.

File: measureme/src/profiling_data.rs (hunk 1 above)
Comment (Mark-Simulacrum) [reply measureme/src/profiling_data.rs]: I imagine that we don't care too much in practice, we just want me to be aware when we make changes -- as it currently stands, we don't really support this too well, but we can either just deal with commits failing or some other way to migrate to a new version. Once we're including measureme in the artifacts rustc CI produces that would be helpful as well.

Comment (wesleywiser) [reply measureme/src/profiling_data.rs]: Ok. We'll be sure to coordinate with you when we bump rustc's dependency version to the one with breaking changes.

File: measureme/src/profiling_data.rs (hunk 2 above)
Comment (michaelwoerister) [reply measureme/src/profiling_data.rs]: You're right. I basically copied this from the string table case, but here a plain EventSink should work just fine.

Comment (michaelwoerister) [reply measureme/src/profiling_data.rs]: I removed the `Arc`.

**Comment by michaelwoerister** [comment]
@wesleywiser I implemented your suggestion and add more test cases. This is ready for review again.

File: summarize/src/analysis.rs [hunk 3]
```diff
@@ -164,3 +164,255 @@ pub fn perform_analysis(data: ProfilingData) -> Results {
...
+        assert_eq!(results.query_data_by_label("q1").invocation_count, 1);
+        assert_eq!(results.query_data_by_label("q2").invocation_count, 1);
+        assert_eq!(results.query_data_by_label("q3").invocation_count, 1);
+    }
+
+    #[test]
+    fn events_with_same_starting_time() {
+        //                      <--e4-->
+        //                      <---e3--->
+        //  <--------e1--------><--------e2-------->
```
Comment (wesleywiser) [inline summarize/src/analysis.rs]: These are super helpful for understanding the tests. Thanks for adding them!

Comment (michaelwoerister) [reply summarize/src/analysis.rs]: Yes, they are super helpful for writing the tests too `:)`

**Pull request merged**